apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/Constants.java

```diff
@@ -18,10 +18,118 @@
+        private final String asyncContextClass;
+
+        // OSGi bundles might only include 'jakarta.servlet' and not 'jakarta.servlet.http' (same for `java.servlet`), hence
+        // this allows to disable instrumentation as it is required for our instrumentation code.
+        private final String osgiClassloaderFilterClass;
+
+        ServletImpl(String servletClass, String filterClass, String requestClass, String responseClass, String filterChainClass,
+                    String asyncContextClass, String osgiClassloaderFilterClass, String servletInputStreamClass, String servletConfigClass) {
+            this.servletClass = servletClass;
+            this.filterClass = filterClass;
+            this.requestClass = requestClass;
+            this.responseClass = responseClass;
+            this.filterChainClass = filterChainClass;
+            this.asyncContextClass = asyncContextClass;
+            this.osgiClassloaderFilterClass = osgiClassloaderFilterClass;
+            this.servletInputStreamClass = servletInputStreamClass;
+            this.servletConfigClass = servletConfigClass;
+        }
+
+        public ElementMatcher.Junction<NamedElement> servletClass() {
+            return named(servletClass);
+        }
+
+        public ElementMatcher.Junction<NamedElement> filterClass() {
+            return named(filterClass);
+        }
+
+        public ElementMatcher.Junction<NamedElement> requestClass() {
+            return named(requestClass);
+        }
+
+        public ElementMatcher.Junction<NamedElement> responseClass() {
+            return named(responseClass);
+        }
+
+        public ElementMatcher.Junction<NamedElement> filterChainClass() {
+            return named(filterChainClass);
+        }
+
+        public ElementMatcher.Junction<NamedElement> asyncContextClass() {
+            return named(asyncContextClass);
+        }
+
+        public ElementMatcher.Junction<TypeDescription> servletInputStreamClass() {
+            return named(servletInputStreamClass);
+        }
+
+        public ElementMatcher.Junction<TypeDescription> servletConfigClass() {
+            return named(servletConfigClass);
+        }
+
+        public ElementMatcher.Junction<ClassLoader> getClassloaderFilter() {
+            // async context class has been introduced in servlet spec 3.0
+            // choice of class name to use for this test does not work as expected across all application servers
+            // for example, 'javax.servlet.annotation.WebServlet' annotation is not working as expected on Payara
+            return CustomElementMatchers.classLoaderCanLoadClass(asyncContextClass)
+                // OSGi bundles might only include 'jakarta.servlet' and not 'jakarta.servlet.http' (same for `java.servlet`), hence
+                // this allows to disable instrumentation as it is required for our instrumentation code.
+                .and(CustomElementMatchers.classLoaderCanLoadClass(osgiClassloaderFilterClass));
```

Comment:
[for reviewer] this is what provides the work-around for OSGi by requiring ability to load a class in `javax.servlet.http.*` or `jakarta.servlet.http.*` packages.

Reply:
This workaround assumes that at least one class on the execution stack is loaded by a class loader that knows both `javax.servlet` and `javax.servlet.http`. So far it seems to be a valid concept, but let's keep that in mind as a possible source for future reports on missing endpoints.